src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -352,7 +352,10 @@ DistributedBlockOutputStream::runWritingJob(DistributedBlockOutputStream::JobRep
                 /// Forward user settings
                 job.local_context = std::make_unique<Context>(context);
 
-                InterpreterInsertQuery interp(query_ast, *job.local_context);
+                /// InterpreterInsertQuery is modifying the AST, but the same AST is also used to insert to remote shards.
```

Comment:
The problem only with INSERT into multiple local shards, i.e. `INSERT INTO FUNCTION cluster('test_cluster_two_shards_localhost', ...)` (AST is not used for remote shards, it uses query string)

So I would say:

```c
/// Copying of the query AST is required to avoid race in case of INSERT into multiple local shards.
/// Since INSERT into local node uses AST, and InterpreterInsertQuery::execute() is modifying it, to resolve tables (in InterpreterInsertQuery::getTable())
```

Reply:
I aggre.